components/x-topic-search/src/TopicSearch.jsx

```diff
@@ -1,95 +1,125 @@
 
-const debounceGetSuggestions = debounce(getSuggestions, 150);
+class TopicSearch extends Component {
+	constructor(props) {
+		super(props);
+
+		this.minSearchLength = props.minSearchLength || 2;
+		this.maxSuggestions = props.maxSuggestions || 5;
+		this.apiUrl = props.apiUrl;
+		this.getSuggestions = debounce(getSuggestions, 150);
+		this.outsideEvents = ['focusout', 'focusin', 'click'];
+		this.handleInputChange = this.handleInputChange.bind(this);
+		this.handleInputClickOrFocus = this.handleInputClickOrFocus.bind(this);
+		this.handleInteractionOutside = this.handleInteractionOutside.bind(this);
+
+		this.state = {
+			searchTerm: '',
+			showResult: false,
+			followedSuggestions: [],
+			unfollowedSuggestions: []
+		};
+	}
+
+	componentDidMount() {
+		this.outsideEvents.forEach(action => {
+			document.body.addEventListener(action, this.handleInteractionOutside);
+		});
+	}
 
-let resultExists = false;
+	componentWillUnmount() {
+		this.outsideEvents.forEach(action => {
+			document.body.removeEventListener(action, this.handleInteractionOutside);
+		});
+	}
+
+	handleInputChange(event) {
+		const searchTerm = event.target.value.trim();
 
-const topicSearchActions = withActions(({ minSearchLength = 2, maxSuggestions = 5, apiUrl, followedTopicIds = [] }) => ({
-	async checkInput(event) {
-		const searchTerm = event.target.value && event.target.value.trim();
+		this.setState({ searchTerm });
 
-		if (searchTerm.length >= minSearchLength) {
-			return debounceGetSuggestions(searchTerm, maxSuggestions, apiUrl, followedTopicIds)
-				.then(result => {
-					resultExists = true;
-					return { showResult: true, result, searchTerm };
+		if (searchTerm.length >= this.minSearchLength) {
+			this.getSuggestions(searchTerm, this.maxSuggestions, this.apiUrl, this.props.followedTopicIds)
+				.then(({ resultsForTerm, followedSuggestions, unfollowedSuggestions }) => {
+					this.setState({
+						resultsForTerm,
+						followedSuggestions,
+						unfollowedSuggestions,
+						showResult: true
+					});
 				})
 				.catch(() => {
-					resultExists = false;
-					return { showResult: false };
+					this.setState({
+						resultsForTerm: null,
+						showResult: false
+					});
 				});
 		} else {
-			resultExists = false;
-			return Promise.resolve({ showResult: false });
+			this.setState({
+				resultsForTerm: null,
+				showResult: false
+			});
 		}
-	},
+	}
 
-	topicFollowed (subjectId) {
-		if (!followedTopicIds.includes(subjectId)) {
-			followedTopicIds.push(subjectId);
+	handleInteractionOutside(event) {
+		if (!this.rootEl.contains(event.target)) {
+			this.setState({
+				showResult: false
+			});
 		}
+	}
 
-		return { followedTopicIds };
-	},
-
-	topicUnfollowed (subjectId) {
-		const targetIdIndex = followedTopicIds.indexOf(subjectId);
-
-		if (targetIdIndex > -1) {
-			followedTopicIds.splice(targetIdIndex, 1);
+	handleInputClickOrFocus() {
+		if (this.state.searchTerm.length >= this.minSearchLength) {
```

Comment:
@dan-searle This line will prevent the bug which it displays results even if there is no searchTerm?

Reply:
Unfortunately not. This only happens when you click into the input. The bug happens when the input already has focus, a request has been made for suggestions, and the response arrives after the results container has been closed (because the input length < 2)

Reply:
Actually, I had to revert that way of fixing it, as it didn't work very well. There's a different way of fixing it now.